sdk/search/Azure.Search.Documents/tests/Samples/FieldBuilder/FieldBuilder.cs

```diff
@@ -0,0 +1,451 @@
+        {
+            if (modelType is null)
+            { throw new ArgumentNullException(nameof(modelType));
+            }
+
+            if (namingPolicy is null)
+            { throw new ArgumentNullException(nameof(namingPolicy));
+            }
+
+            ArgumentException FailOnNonObjectDataType()
+            {
+                string errorMessage =
+                    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " +
+                    "class or struct with public properties.";
+
+                throw new ArgumentException(errorMessage, nameof(modelType));
+            }
+
+            if (ObjectInfo.TryGet(modelType, out ObjectInfo info))
+            {
+                if (info.Properties.Length == 0)
+                {
+                    throw FailOnNonObjectDataType();
+                }
+
+                // Use Stack to avoid a dependency on ImmutableStack for now.
+                return BuildForTypeRecursive(modelType, info, namingPolicy, new Stack<Type>(new[] { modelType }));
+            }
+
+            throw FailOnNonObjectDataType();
+        }
+
+        private static IList<SearchField> BuildForTypeRecursive(
+            Type modelType,
+            ObjectInfo info,
+            JsonNamingPolicy namingPolicy,
+            Stack<Type> processedTypes)
+        {
+            SearchField BuildField(PropertyInfo prop)
+            {
+                static bool ShouldIgnore(Attribute attribute) =>
+                    attribute is JsonIgnoreAttribute || attribute is FieldBuilderIgnoreAttribute;
+
+                IList<Attribute> attributes = prop.GetCustomAttributes(true).Cast<Attribute>().ToArray();
+                if (attributes.Any(ShouldIgnore))
+                {
+                    return null;
+                }
+
+                SearchField CreateComplexField(SearchFieldDataType dataType, Type underlyingClrType, ObjectInfo info)
+                {
+                    try
+                    {
+                        if (processedTypes.Contains(underlyingClrType))
+                        {
+                            // Skip recursive types.
+                            return null;
+                        }
+
+                        processedTypes.Push(underlyingClrType);
+                        IList<SearchField> subFields =
+                            BuildForTypeRecursive(underlyingClrType, info, namingPolicy, processedTypes);
+
+                        string fieldName = namingPolicy.ConvertName(prop.Name);
+
+                        SearchField field = new SearchField(fieldName, dataType);
+                        foreach (SearchField subField in subFields)
+                        {
+                            field.Fields.Add(subField);
+                        }
+
+                        return field;
+                    }
+                    finally
+                    {
+                        processedTypes.Pop();
```

Comment:
Do we need to scope the `try` to right after we push the type?  Would this pop somebody else off the stack if we had a recursive type?

Reply:
While this was meant to be a straight-up port with not additional time spent, I agree this is a bug I should fix now.

Reply:
Done